Unify returns result rather than raising an exception

samsa1 opened the pull request:
I have noticed that almost all calls to `unify` are directly wrapped inside an exception handler. The cases where we don't have a direct exception handler are the exception that most of the time just expect the unification to never fail.

As such I propose to replace the exception by returning a result. As such the type-checker will help prevent in any future evolution any uncaught exception.

In order to stay as close as possible to the actual semantic I introduced a `unify_ex` (and a `unify_var_ex`) that returns an exception for the few cases where we didn't have a direct exception handler. A futur improvement might be to remove this function from the interface of ctype.ml and adding assertions in places we expect that the unification will never fail.


This PR is reviewed on top of #14422

samsa1 committed:
Changed the interface of unify to return a result

samsa1 committed:
Changed the interface of unify to return a result

samsa1 committed:
Moved a match outside of an if in typetexp

samsa1 committed:
Handled comments

samsa1 committed:
Changed the interface of unify_var and unify_gadt to return a result rather than an exception

samsa1 committed:
Changed the interface of unify_var and unify_gadt to return a result rather than an exception

gasche commented:
I would prefer `_exn` as the exception prefix, as that convention already exists in some parts of the compiler codebase.

johnyob commented:
> I would prefer `_exn` as the exception prefix, as that convention already exists in some parts of the compiler codebase.

Was literally writing the same comment 😅

johnyob commented on typing/ctype.mli:
```diff
@@ -288,20 +288,26 @@ val extract_concrete_typedecl:
+val unify_ex: Env.t -> type_expr -> type_expr -> unit
         (* Unify the two types given. Raise [Unify] if not possible. *)
+val unify: Env.t -> type_expr -> type_expr ->
+        (unit, Errortrace.unification_error) Result.t
```
nit: prefer `result` over `Result.t`

I would cite that this is consistent with the compiler codebase, but the compiler uses both `result` and `Result.t` haphazardly ://

samsa1 replied:
A side question is do we want to define a local allias for the type `('a, Errortrace.unification_error) Result.t` such as `' unification_result` ?

johnyob replied:
This would also be a nice pattern to use within the compiler :) e.g. `'a Errortrace.unification_result`

gasche commented on utils/misc.ml:
```diff
@@ -141,6 +141,19 @@ module Stdlib = struct
+        ([], []) -> Ok ()
+      | (a1::l1, a2::l2) ->
+        Result.bind (f a1 a2) (fun () -> iter2_result f l1 l2)
+      | (_, _) -> raise (Invalid_argument "iter2_result")
```
I feel slightly nervous about the fact that these functions are not self-recursive on the nose: they are tail-recursive but go via an indirect call by the higher-order function `Result.bind`, so they are not going to be tail-recursive with js_of_ocaml. This may or may not be something we care about depending on the usage and circumstances (this is not a stdlib addition), but in this case it's trivial to make this issue go away by inlining the call to `Result.bind`. Could you do this?

johnyob replied:
Alternatively, the `iter` and `iter2` functions could be used with a local exception that is raised and caught on the error path? Not entirely sure what this would look like on `js_of_ocaml`, but has the benefit of reusing the `Stdlib` `iter` and `iter2`?

samsa1 replied:
I feel more inclined to follow the proposition from @gasche but if you feel the second option is better I can go this way. What are you preferences between both options ?

gasche commented on toplevel/topprinters.ml:
```diff
@@ -63,14 +63,16 @@ let match_simple_printer_type env ty ~is_old_style =
+          (make_printer_type ty_arg)
+          (Ctype.instance ty)
+      in
+      unify_res, ty_arg
```
Conceptually this is a bit ugly because the `ty_arg` argument makes no sense in the error case. I would prefer:

```ocaml
let open Result.Syntax in
let ty_arg = Ctype.newvar () in
let+ () = Ctype.unify ... in
ty_arg
```

johnyob commented on typing/typeclass.ml:
```diff
@@ -355,11 +357,8 @@ and class_signature virt env pcsig self_scope loc =
-  with Ctype.Unify _ ->
-    raise(Error(sty.ptyp_loc, env, Pattern_type_clash self_type))
-  end;
+  if Result.is_error (Ctype.unify env self_type sign.csig_self) then
+    raise(Error(sty.ptyp_loc, env, Pattern_type_clash self_type));
```
My first reaction to this pattern was 'there must be a `Result.ok_or_else` combinator for this', but I was wrong

johnyob replied:
Also this PR uses `match ... with Ok () -> () | Error err -> raise ...` a lot, why switch it up here?

samsa1 replied:
I feel the `if Result.is_error ...` to be more readable than having an auxiliairy function for this but defined a new function in `Result` for the `match ... with Ok () -> () | Error err -> raise ...` case (called `Result.ok_or_else` but this name can easily be changed)

gasche commented on typing/typecore.ml:
```diff
@@ -481,12 +481,13 @@ let type_continuation_pat env expected_ty sp =
+  | Error err ->
       raise(Error(loc, env, Expr_type_clash(err, None, None)))
-  | Tags(l1,l2) ->
+  | exception Tags(l1,l2) ->
```
Ah, of course `unify` raises another possible exception, `Tags ...`, which is undocumented in the .mli, has an unclear status (it has its own error handler registered by the compiler... but is also rewrapped as another error in some cases), and remains raised as an exception rather than being part of the `Error` cases.

I think that the choice that you made to leave it as an exception is the correct one, and it's not your fault that it was never documented in the first place, but could you mention it in ctype.mli?

samsa1 replied:
I added a comment stating that `Tags` can be raised without more details. Should I try to give a more details ?

johnyob commented on typing/typecore.ml:
```diff
@@ -2934,8 +2939,9 @@ let collect_unknown_apply_args env funct ty_fun0 rev_args sargs =
                   Warnings.Ignored_extra_argument;
-              unify env ty_fun
-                (newty (Tarrow(lbl,ty_param,ty_res,commu_var ())));
+              Result.get_ok
```
Perhaps we should use `set_type_desc` here instead of `unify`?

samsa1 replied:
This change here is wrong because the unification changes the levels. As such we would have an incorrect level on `ty_param` and `ty_res`. Creating them with the correct level will remove this issue but I propose to do this in a separate PR as it is a non-trivial change

gasche reviewed (commented):
After looking at the patch, I believe that it is correct but I am also not fully convinced that the change is necessary / obviously an improvement. The fact that `unify` returned `unit` before made some patterns convenient, for example

```ocaml
try unify <args>
with Unify err -> <error case>
```

becomes the more boilerplate-y

```ocaml
match unify <args> with
| Ok () -> ()
| Error err -> <error case>
```

This is not a negative vote, but let's say I would be waiting for someone else to express clear support.

johnyob commented on typing/typecore.ml:
```diff
@@ -5377,10 +5394,7 @@ and type_function
-              try unify env (type_option ty_default) ty_arg_mono
-              with Unify _ -> assert false;
-            end;
+            Result.get_ok (unify env (type_option ty_default) ty_arg_mono);
```
Shouldn't this be `assert (Result.is_ok ...)`?

gasche replied:
As an occasionally `-noassert`-paranoid person, I would not recommend sticking important side-effects inside assertions.

johnyob replied:
My problem: the `Invalid_argument` raised by `Result.get_ok` will be less helpful than the original `assert false` in diagnosing any issue with the (undocumented) invariant here

samsa1 replied:
Would a `if Result.is_error ... then failwith ...` be a good compromise then ?

johnyob commented:
Agreeing with @gasche here, the patch looks correct. I do believe the change is an improvement (the use of a result here forces me to handle the error case, which I may forget about w/out LSP to quickly look at the documentation). 

However, I also agree with @gasche that the translation of
```ocaml
try unify <args>
with Unify err -> <error case>
```
into the more boilerplate-y
```ocaml
match unify <args> with
| Ok () -> ()
| Error err -> <error case>
```
is unpleasant. 

In Rust, you'd probably write something along the lines of `Result.get_ok_or (unify <args>) ~else:(fun err -> <error case>)` (with OCaml naming).

gasche commented:
We discussed this again during the type-system meeting today. @samsa1 made the point that, even if it does not clearly improve the codebase, it does make it more approachable with newcomers. I find this a reasonable argument and will be happy to merge once the CI is green again.

garrigue commented:
@gasche , you are a bit fast to approve.
Shouldn't the name exception version be changed to `unify_exn` at least ?
For the rest, I was planning to give it a look, but I suppose I can do it later (it is not going to break anything anyway).

garrigue commented:
BTW, I found the place I was talking about.
This is `expand_head_unif`, in `ctype`.
This is one of the rare functions raising `Unify`, and it might benefit from the same change.

gasche commented:
> Shouldn't the name exception version be changed to `unify_exn` at least ?

I agree. @samsa1, if/when you have time, it should be `unify_exn` and not `unify_ex`. Other than that I think that the weak consensus to merge still stands, and @garrigue's study has not surfaced any pressing change that should be made to the PR, so I would still be in favor of approving and merging if you rebase the PR.

johnyob commented:
I was just looking at exceptions raised by `unify` (in the context of #14538) and was reminded how much of an improvement this PR would be 😅

garrigue commented:
We did agree to merge it at the meeting two or three months ago. I think we are just waiting for a rebase.

samsa1 commented:
Sorry for the long sleep on this PR. I tried to make a pass on all the comments above and all of them have been handled in some way but I consider 2 things to point out :
- I added a utility function `Result.ok_or_else : ('a, 'b) result -> ('b -> 'a) -> 'a` to prevent most of the matches but I am fully open to better naming propositions (or a let-binding syntax).
- I haven't changed  the interface of `expand_head_unif` because it is internal to ctype and most use cases rely on the exception (5 out of 7) but I changed `expand_head_trace` that will benefit from this change.
